File: nodecore-cli/src/main/java/nodecore/cli/ProgramOptions.java
// VeriBlock NodeCore CLI
// Copyright 2017-2019 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.

package nodecore.cli;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.CommandLineParser;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.Option;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;

import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ProgramOptions {
    private String _configPath;
    private List<String> _scripts = new ArrayList<>();
    private Properties _properties = new Properties();

    private String _connect;

    public ProgramOptions() {
        resetToDefaults();
    }

    public boolean parse(String[] args) {
        Option configFileOption = Option.builder("c")
                .argName("path")
                .hasArg()
                .required(false)
                .desc("The configuration file location")
                .longOpt("config")
                .build();

        Option paramOption = Option.builder("D")
                .required(false)
                .desc("Specify a config override in key=value form")
                .argName("property=value")
                .numberOfArgs(2)
                .valueSeparator()
                .build();

        Option connectOption = Option.builder("connect")
                .argName("address")
                .hasArg()
                .required(false)
                .desc("Specify a node to connect to on startup")
                .longOpt("connect")
                .build();

        Options options = new Options();
        options.addOption(configFileOption);
        options.addOption(paramOption);
        options.addOption(connectOption);

        CommandLineParser parser = new DefaultParser();
        try {
            CommandLine commandLine = parser.parse(options, args);

            _properties = commandLine.getOptionProperties("D");

            if (commandLine.hasOption("c"))
                _configPath = commandLine.getOptionValue('c');

            if (commandLine.hasOption("connect"))
                _connect = commandLine.getOptionValue("connect");

            return true;
        } catch (ParseException e) {
            System.out.println(e.getMessage());
            return false;
        }
    }

    public String getConfigPath() {
        return _configPath;
    }

    public void resetToDefaults() {
        _properties.clear();
        _configPath = "nodecore-cli.properties";
    }

    public void removeProperty(String name) {
        _properties.remove(name);
    }

    public String getProperty(String name) {
        return _properties.getProperty(name);
    }

    public String getConnect() {
        return _connect;
    }
}

File: nodecore-cli/src/main/java/nodecore/cli/ProgramOptions.kt
// VeriBlock NodeCore CLI
// Copyright 2017-2021 Xenios SEZC
// All rights reserved.
// https://www.veriblock.org
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php.
package nodecore.cli

import org.apache.commons.cli.CommandLineParser
import org.apache.commons.cli.DefaultParser
import org.apache.commons.cli.Option
import org.apache.commons.cli.Options
import org.apache.commons.cli.ParseException
import java.util.*

class ProgramOptions {
    lateinit var configPath: String
        private set

    var connect: String? = null

    private var properties = Properties()

    init {
        resetToDefaults()
    }

    fun parse(args: Array<String>): Boolean {
        val configFileOption = Option.builder("c")
            .argName("path")
            .hasArg()
            .required(false)
            .desc("The configuration file location")
            .longOpt("config")
            .build()
        val paramOption = Option.builder("D")
            .required(false)
            .desc("Specify a config override in key=value form")
            .argName("property=value")
            .numberOfArgs(2)
            .valueSeparator()
            .build()
        val connectOption = Option.builder("connect")
            .argName("address")
            .hasArg()
            .required(false)
            .desc("Specify a node to connect to on startup")
            .longOpt("connect")
            .build()
        val options = Options()
        options.addOption(configFileOption)
        options.addOption(paramOption)
        options.addOption(connectOption)
        val parser: CommandLineParser = DefaultParser()
        return try {
            val commandLine = parser.parse(options, args)
            properties = commandLine.getOptionProperties("D")
            if (commandLine.hasOption("c")) {
                configPath = commandLine.getOptionValue('c')
            }
            if (commandLine.hasOption("connect")) {
                connect = commandLine.getOptionValue("connect")
            }
            true
        } catch (e: ParseException) {
            println(e.message)
            false
        }
    }

    fun resetToDefaults() {
        properties.clear()
        configPath = "nodecore-cli.properties"
    }

    fun removeProperty(name: String) {
        properties.remove(name)
    }

    fun getProperty(name: String): String? {
        return properties.getProperty(name)
    }
}
